### ISSUE TITLE
False positive "trait bound not satisfied" (E0277) for multi-step implementation involving `impl` and a trait bound on generic type in different crates

### ISSUE THREAD

**Issue by Kyykek** [intellij-rust/intellij-rust#10615]
## Environment

* **IntelliJ Rust plugin version:** 0.4.196.5423-231
* **Rust toolchain version:** 1.70.0 (90c541806 2023-05-31) x86_64-pc-windows-msvc
* **IDE name and version:** IntelliJ IDEA 2023.1 Ultimate Edition (IU-231.8109.175)
* **Operating system:** Windows 10.0
* **Macro expansion:** enabled

## Problem description
The code below compiles, but the plugin gives a false E0277.

## Steps to reproduce
Crate `some-library` (lib.rs):
```rust
pub struct GenericType<T>(T);

pub trait BaseTrait {}

// `T: BaseTrait` implies `GenericType<T>: BaseTrait`
impl<T: BaseTrait> BaseTrait for GenericType<T> {}
```
Crate `bug-demo` (lib.rs):
```rust
use some_library::{BaseTrait, GenericType};

pub trait BoundTrait: BaseTrait {}

// the trait bound `GenericType<T>: BaseTrait` is not satisfied [E0277]
// the trait `BaseTrait` is not implemented for `GenericType<T>`
impl<T: BoundTrait> BoundTrait for GenericType<T> {}
```

### Why it's valid
- `BoundTrait: BaseTrait`  (trait bound in `bug-demo`)
	- This can be read as `T: BoundTrait` requires `T: BaseTrait`
	- That is another way of saying `T: BoundTrait` implies `T: BaseTrait`
- `T: BaseTrait` implies `GenericType<T>: BaseTrait`   (`impl` in `some-library`)

Therefore, `T: BoundTrait` transitively implies `GenericType<T>: BaseTrait`.
The erroring impl does require `T: BoundTrait`, so `GenericType<T>: BaseTrait` is actually satisfied.

The error disappears when `BaseTrait` and `BoundTrait` are in the same crate. Note that I only tested this with local crates.

**Comment by Maksim20023** [comment]
Hi! Thank you for reporting the issue.
Unfortunately, I was unable to reproduce the issue based on the information in description, and it seems that this "false positive {E0277]" issue was already fixed: [GitHub pull 9639](https://github.com/intellij-rust/intellij-rust/pull/9639)
As a general troubleshooting step, please try to [update](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#update) Rust plugin to actual [0.4.197.5401-231](https://plugins.jetbrains.com/plugin/8182-rust/versions)  and [invalidating caches](https://www.jetbrains.com/help/idea/invalidate-caches.html).
If the issue still persists, it would be great if you could send me your crates files and [logs](https://www.jetbrains.com/help/idea/remote-development-troubleshooting.html#logs_UI), so i can take a look at what is going on. Please keep in mind that `idea.log` might contain sensitive information (e.g local paths). You can upload your logs and snapshot to our https://uploads.jetbrains.com. Uploaded files will only be accessible to JetBrains employees, and we guarantee the privacy of the data

**Comment by Kyykek** [comment]
Interesting, I was

**Comment by Maksim20023** [comment]
Thanks for your project. It's very strange, but I still can't reproduce this issue. I tried doing it on both Windows and a clean Linux setup. It seems that the issue may appear in the project with multiple crates which aren't combined into a [workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html). Could you please try to place your project into [Cargo workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html)?

**Comment by Kyykek** [comment]
Welp, here I am writing Rust for years while never having used a workspace. I've seen it quite a few times, but never felt the need for it since it works fine without it; guess there is always something to learn.

The workspace does "fix" the issue, but since this is perfectly working code even without one, i would still consider this a bug.

**Comment by Kyykek** [comment]
This also makes me wonder about why many inspections currently cannot be suppressed when using the plugin. This is not the first time I have encountered False positives and it wouldn't be a big deal if it was possible to just suppress them, since the external linter can do most of the heavy lifting. 
I'm aware of the plugin API's reputation, so I'm curious, are there technical limitations/some other reasons why this would be hard to implement?

**Comment by Maksim20023** [comment]
> Welp, here I am writing Rust for years while never having used a workspace. I've seen it quite a few times, but never felt the need for it since it works just fine without it; guess there is always something to learn.
> 
> The workspace does "fix" the issue, but since this is perfectly working code even without one, i would still consider this a bug.

Using a workspace is not mandatory. However, it's recommended when working with multiple crates in a single project. Unfortunately, I currently have no information about any planned changes or fixes.

> This also makes me wonder about why many inspections currently cannot be suppressed when using the plugin. This is not the first time I have encountered false positives and it wouldn't be a big deal if it was possible to just suppress them, since the external linter can do most of the heavy lifting.
> I'm aware of the plugin API's reputation, so I'm curious, are there technical limitations/some other reasons why this would be hard to implement?

False positive warnings can be caused by various reasons. If you encounter such a warning, feel free to add a comment with a description to the existing ticket, or create a new one. Your report could be very valuable to our development team.

**Comment by Kyykek** [comment]
> I just wonder why suppression doesn't work as documented [here](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-analysis.html#suppress)?
Suppression has always been shaky it comes to suppressing errors with types in statically typed languages because they are needed in other parts (or at least that's what i've been told, same thing for syntax in all languages). 
False positives happen, especially when you're basically implementing an entirely seperate linter from the official language utilities. Intellij really needs a "suppress everything here, yes i am sure, it's ok if other analysis fails because of this" option, as false positives that cannot be turned off heavily degrade the usefulness of error checking as a whole.
